File: Data-Structures/Tree/Binary-Tree/SymmetryBTree.cs
// C# program to check is binary  
// tree is symmetric or not  
using System;  
  
class Node  
{  
    public int key;  
    public Node left, right;  
      
    public Node(int item)  
    {  
        key = item;  
        left = right = null;  
    }  
}  
      
class GFG  
{  
    Node root;  
      
    // returns true if trees with roots 
    // as root1 and root2 are mirror  
    Boolean isMirror(Node node1,  
                     Node node2)  
    {  
        // if both trees are empty,  
        // then they are mirror image  
        if (node1 == null && node2 == null)  
            return true;  
      
        // For two trees to be mirror images,  
        // the following three conditions must be true  
        // 1 - Their root node's key must be same  
        // 2 - left subtree of left tree and right subtree  
        //       of right tree have to be mirror images  
        // 3 - right subtree of left tree and left subtree  
        //       of right tree have to be mirror images  
        if (node1 != null && node2 != null &&  
            node1.key == node2.key)  
            return (isMirror(node1.left, node2.right) &&  
                    isMirror(node1.right, node2.left));  
      
        // if neither of the above conditions  
        // is true then root1 and root2 are 
        // mirror images  
        return false;  
    }  
      
    // returns true if the tree is symmetric  
    // i.e mirror image of itself  
    Boolean isSymmetric(Node node)  
    {  
        // check if tree is mirror of itself  
        return isMirror(root, root);  
    }  
      
    // Driver Code  
    static public void Main(String[] args)  
    {  
        GFG tree = new GFG();  
        tree.root = new Node(1);  
        tree.root.left = new Node(2);  
        tree.root.right = new Node(2);  
        tree.root.left.left = new Node(3);  
        tree.root.left.right = new Node(4);  
        tree.root.right.left = new Node(4);  
        tree.root.right.right = new Node(3);  
        Boolean output = tree.isSymmetric(tree.root);  
        if (output == true)  
            Console.WriteLine("1");  
        else
            Console.WriteLine("0");  
    }  
}  

File: Data-Structures/Tree/Binary-Tree/SymmetryBTree.py
# Node structure 
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        """"
        Creates a new tree node
        :param val: Value of the node
        :param left: Node on the left
        :param right: Node on the right
        """
        self.val = val
        self.left = left
        self.right = right
    

def isMirror(root1, root2):
    """"
    Returns True if trees with roots as root1 and root 2  are mirror 
    :param root1: TreeNode
    :param root2: TreeNode
    """
    # If both trees are empty, then they are mirror images 
    if root1 is None and root2 is None: 
        return True

    """ For two trees to be mirror images, the following three 
        conditions must be true 
        1 - Their root node's key must be same 
        2 - left subtree of left tree and right subtree 
        of the right tree have to be mirror images 
        3 - right subtree of left tree and left subtree 
        of right tree have to be mirror images 
    """
    if (root1 is not None and root2 is not None): 
            if root1.val == root2.val: 
                return (isMirror(root1.left, root2.right)and
                isMirror(root1.right, root2.left)) 

    # If neither of the above conditions is true then root1 
    # and root2 are not mirror images 
    return False

def isSymmetric(root: TreeNode):
    """
    Checks if the tree is mirror of itself
    :param root: Tree Node
    """
    # Check if tree is mirror of itself 
    return isMirror(root, root)


# Driver Program 
# Let's construct the tree show in the above figure 
root = TreeNode(1) 
root.left = TreeNode(2) 
root.right = TreeNode(2) 
root.left.left = TreeNode(3) 
root.left.right = TreeNode(4) 
root.right.left = TreeNode(4) 
root.right.right = TreeNode(3) 
print("1" if isSymmetric(root) == True else "0")

# This code is contributed by Nikhil Kumar Singh(nickzuck_007) 
